From bca459782e8c847cb2e1f2d9fc775089256acf3c Mon Sep 17 00:00:00 2001
From: Vaibhav Bedia <vaibhav.bedia@ti.com>
Date: Wed, 14 Mar 2012 22:53:59 +0530
Subject: [PATCH 198/609] ARM: OMAP: AM33XX: PM: Don't rely on M3 for A8
 Standby

Original commit: 11ddaf2ff9d4be13cebd0ec7fbc4b83a022b915e

  From git://arago-project.org/git/projects/linux-am33x.git
  And ti-sdk-am335x-evm-05.05.00.00-Linux-x86-Install image

Relying on M3 to have A8 assert standby is not really necessary.
It also needs to unnecessary complexity in the recovery path.
So, we write to the CLKCTRL register when A8 is sure that it is
going to enter the low power state.

Signed-off-by: Vaibhav Bedia <vaibhav.bedia@ti.com>
---
 arch/arm/mach-omap2/pm33xx.c |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/pm33xx.c b/arch/arm/mach-omap2/pm33xx.c
index 47b907f..70bcb42 100644
--- a/arch/arm/mach-omap2/pm33xx.c
+++ b/arch/arm/mach-omap2/pm33xx.c
@@ -37,6 +37,7 @@
 #include <asm/sizes.h>
 
 #include "pm.h"
+#include "cm33xx.h"
 #include "pm33xx.h"
 #include "control.h"
 #include "clockdomain.h"
@@ -45,6 +46,8 @@
 void (*am33xx_do_wfi_sram)(void);
 
 #define DS_MODE		DS0_ID	/* DS0/1_ID */
+#define MODULE_DISABLE	0x0
+#define MODULE_ENABLE	0x2
 
 #ifdef CONFIG_SUSPEND
 
@@ -167,8 +170,12 @@ static int am33xx_pm_suspend(void)
 	else
 		pr_err("Could not program GFX to low power state\n");
 
+	writel(0x0, AM33XX_CM_MPU_MPU_CLKCTRL);
+
 	ret = cpu_suspend(0, am33xx_do_sram_idle);
 
+	writel(0x2, AM33XX_CM_MPU_MPU_CLKCTRL);
+
 	if (gfx_pwrdm) {
 		state = pwrdm_read_pwrst(gfx_pwrdm);
 		if (state != PWRDM_POWER_OFF)
-- 
1.7.5.4

